Update `update` wiki page (#1008)
iotaledger/identity.rs: documentation/docs/concepts/decentralized_identifiers/update.mdx

description: How DID Documents can be manipulated and how updates should be published
image: /img/Identity_icon.png
keywords:
  - Documents
  - DID
  - Tangle
  - Update
  - Publish
---

import account_manipulate_did_rs from "!!raw-loader!../../../../examples_legacy/account/manipulate_did.rs";
import update_did_rust from "!!raw-loader!../../../../examples/0_basic/1_update_did.rs";
import update_did_node from "!!raw-loader!../../../../bindings/wasm/examples/src/ex1_update_did.ts";
import CodeSnippet from "../../../src/components/CodeSnippetComponent";
import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";
import CodeBlock from "@theme/CodeBlock";

> DID Documents can be extended by adding [verification methods](https://www.w3.org/TR/did-core/#verification-methods), [Services](https://www.w3.org/TR/did-core/#services) and custom properties.

Reviewer: Link to the sections which then should link out to the specification. Also this article does not cover custom properties other than a single sentence under `Services`, should it have a proper section?

```suggestion
DID Documents can be extended by adding [verification methods](#verification-methods), [services](#services) and custom properties.
```

A verification method adds public keys, which can be used to digitally sign things like a DID message or a verifiable credential, while a service can provide metadata around the identity via URIs.

### Verification Methods

> As demonstrated by the [example](#example) below, the Iota identity framework offers easy-to-use methods for adding verification methods.

Reviewer: ```suggestion
As demonstrated by the [example](#example) below, the IOTA Identity framework offers easy-to-use methods for adding [verification methods](https://www.w3.org/TR/did-core/#verification-methods).
```

Upon creating a verification method using the identity framework, specifying the `MethodScope` option will result in an embedded verification method. Leaving that option unset will create the verification method as
a map entry of the `verificationMethod` property. Verification relationships can be added afterwards using references.

:::note

> Updates to the DID Document are done through a state transition of the [Alias Output](../../specs/did/iota_did_method_spec#alias-output). The State Controller does not have to be listed as a Capability Invocation verification method.

Reviewer: The capability invocation reference is only relevant to previous users, it's more accurate to say it doesn't have to be included at all, if we keep that sentence.

```suggestion
Updates to the DID Document are done through a state transition of the [Alias Output](../../specs/did/iota_did_method_spec#alias-output) by its state controller. The public key or address of the state controller does not need to be a verification method in the DID Document, since it is defined in the containing Alias Output.
```

### Services

[Services](https://www.w3.org/TR/did-core/#services) allow adding other ways of communicating with the DID subject. An endpoint included in the DID Document can offer a way of reaching services for different purposes
like authentication, communicating, and discovery.

The following properties can be specified for a service:

- **id**: a [DID URL](https://www.w3.org/TR/did-core/#did-url-syntax) for referencing the service in the DID document.
  It can be specified by setting the [fragment](https://www.w3.org/TR/did-core/#fragment).
- **type**: a string used to maximize interoperability between services. The framework does not perform any checks on the content of this string.
- **serviceEndpoint**: a URL that points to the service endpoint.

## Example

The following example demonstrates managing verification methods and services to a DID Document.

<CodeSnippet
  nodeContent={update_did_node}
  rustContent={update_did_rust}
  nodeGithubLink="https://github.com/iotaledger/identity.rs/blob/dev/bindings/wasm/examples/src/ex1_update_did.ts"
  rustGithubLink="https://github.com/iotaledger/identity.rs/blob/dev/examples/0_basic/1_update_did.rs"
/>

### Creating Identity

> The Example above starts by [creating an identity](./create.mdx).

Reviewer: ```suggestion
The example above starts by [creating an identity](./create.mdx).
```

This method is used to sign the DID Document for publication to the Tangle.
The signature proves that the publisher of the document is in control over the capability invocation keys and is allowed to create, update or delete the DID Document.

Any future updates to the DID Document in this example will be signed using this verification method. The Account will automatically sign each update with this method so individual
updates don't have to be explicitly signed.

Furthermore, it's possible to rotate a capability Invocation key. In this case, the Account will sign next update with a key which was valid in the previous state of the DID Document. Afterwards it will
 use the first (oldest) of the remaining capability invocation keys as a default signing method.
Other capability invocation keys can still be explicitly specified to sign an update. These can be set in `PublishOptions`.

Note that the Account does not allow removing all capability invocation keys.
// Create a new DID in an Alias Output for us to modify.
let (_, did): (Address, StardustDID) = create_did(&client, &mut secret_manager).await?;

> ### Adding Verification Methods

Reviewer: The example should be updated to demonstrate adding a verification method, since currently that operation is hidden by calling `create_did`, since it's no longer self-contained.

> This will creates and publishes an Alias Output containing a DID Document with one verification method.

Reviewer: ```suggestion
This creates and publishes an Alias Output containing a DID Document with one verification method.
```

        "publicKeyMultibase": "z8ogKEagUH8swGf9vJNwsLQk79HSm2rkbQjfsMuGQqBjG"
      }
    ],
    "authentication": ["did:iota:rms:0x7408d40ceaf44b83ad03b0e05f5a5fe21adce0b47a133f91657efe4e41619a06#key-1"]
  },
  "meta": {
    "created": "2022-09-11T19:21:47Z",
    "updated": "2022-09-11T19:21:47Z"
  }
}
```

Notice that these changes to the document are not published yet. This will be done after adding the service.

### Adding a Service

<Tabs groupId="code-examples">
<TabItem value="Rust" label="Rust">

```rust
// Add a new Service.
let service: StardustService = Service::from_json_value(json!({
    "id": document.id().to_url().join("#linked-domain")?,
    "type": "LinkedDomains",
    "serviceEndpoint": "https://iota.org/"
}))?;
assert!(document.insert_service(service));
document.metadata.updated = Some(Timestamp::now_utc());
```

</TabItem>
<TabItem value="node" label="Node.js">

```js
// Add a new Service.
const service: StardustService = new StardustService({
  id: did.join("#linked-domain"),
  type: "LinkedDomains",
  serviceEndpoint: "https://iota.org/"
});
document.insertService(service);
document.setMetadataUpdated(Timestamp.nowUTC());
```

</TabItem>
</Tabs>

> In JavaScript, the `serviceEndpoint` property type is a string, this must be a valid URL, otherwise an error will be thrown.

Reviewer: Both Rust and Node.js use the JSON constructor, so this doesn't apply. Also we support sets and maps of endpoints now, so it's not entirely accurate, maybe linking out would be better.

```suggestion
```

            "type":"MyCustomService",
            "serviceEndpoint":"https://example.com/"
         }
      ]
   },
   "meta":{
      "created":"2022-04-13T09:27:48Z",
      "updated":"2022-04-13T09:28:34Z"
   }
  "doc": {
    "id": "did:iota:rms:0x7408d40ceaf44b83ad03b0e05f5a5fe21adce0b47a133f91657efe4e41619a06",
    "verificationMethod": [
      {
        "id": "did:iota:rms:0x7408d40ceaf44b83ad03b0e05f5a5fe21adce0b47a133f91657efe4e41619a06#key-1",
        "controller": "did:iota:rms:0x7408d40ceaf44b83ad03b0e05f5a5fe21adce0b47a133f91657efe4e41619a06",
        "type": "Ed25519VerificationKey2018",
        "publicKeyMultibase": "z8ogKEagUH8swGf9vJNwsLQk79HSm2rkbQjfsMuGQqBjG"
      }
    ],
    "authentication": ["did:iota:rms:0x7408d40ceaf44b83ad03b0e05f5a5fe21adce0b47a133f91657efe4e41619a06#key-1"],
    "service": [
      {
        "id": "did:iota:rms:0x7408d40ceaf44b83ad03b0e05f5a5fe21adce0b47a133f91657efe4e41619a06#linked-domain",
        "type": "LinkedDomains",
        "serviceEndpoint": "https://iota.org/"
      }
    ]
  },
  "meta": {
    "created": "2022-09-11T19:21:47Z",
    "updated": "2022-09-11T19:23:24Z"
  }
}
```

> ### Removing a Verification Method

Reviewer: Again, update the example to demonstrate how to remove a verification method.

> Upon creating a verification method using the identity framework, specifying the `MethodScope` option will result in an embedded verification method. Leaving that option unset will create the verification method as

Reviewer: The "embedded or referenced" URL above in the sentence before this (I can't highlight it) is a broken link, the reference changed.

New link:
https://www.w3.org/TR/did-core/#referring-to-verification-methods
